build/cloudbuild-tf-apply.yaml

```diff
@@ -14,31 +14,51 @@
 
 timeout: 1200s
 steps:
-- name: gcr.io/$PROJECT_ID/terraform
+- id: "set SA"
+  name: gcr.io/${PROJECT_ID}/terraform
   entrypoint: /bin/bash
   args:
   - -c
   - |
     gcloud config set auth/impersonate_service_account ${_TF_SA_EMAIL}
-- name: gcr.io/$PROJECT_ID/terraform
+
+# [START tf-init]
+- id: 'tf init'
+  name: gcr.io/${PROJECT_ID}/terraform
+  entrypoint: /bin/bash
   args:
-  - init
-  - -input=false
-- name: gcr.io/$PROJECT_ID/terraform
+  - -c
+  - |
+      ./tf-wrapper.sh init ${BRANCH_NAME}
+
+# [START tf-validate]
+- id: 'tf validate'
+  name: gcr.io/${PROJECT_ID}/terraform
+  entrypoint: /bin/bash
   args:
-  - validate
-- name: gcr.io/$PROJECT_ID/terraform
+  - -c
+  - |
+      ./tf-wrapper.sh validate ${BRANCH_NAME} ${_POLICY_REPO}
+
+# [START tf-plan]
+- id: 'tf plan'
+  name: gcr.io/${PROJECT_ID}/terraform
+  entrypoint: /bin/bash
   args:
-  - plan
-  - -input=false
-  - -out=tfplan
-- name: gcr.io/$PROJECT_ID/terraform
+  - -c
+  - |
+      ./tf-wrapper.sh plan ${BRANCH_NAME}
+
+# [START tf-apply]
+- id: 'tf apply'
+  name: gcr.io/${PROJECT_ID}/terraform
+  entrypoint: /bin/bash
   args:
-  - apply
-  - -auto-approve
-  - -input=false
-  - tfplan
+  - -c
+  - |
+      ./tf-wrapper.sh apply ${BRANCH_NAME}
+
 artifacts:
   objects:
-    location: 'gs://${_ARTIFACT_BUCKET_NAME}/terraform/cloudbuild/apply/$BUILD_ID'
-    paths: ['cloudbuild-tf-apply.yaml', 'tfplan']
+    location: 'gs://${_ARTIFACT_BUCKET_NAME}/terraform/cloudbuild/apply/${BUILD_ID}'
+    paths: ['cloudbuild-tf-apply.yaml', 'envs/${BRANCH_NAME}/${BRANCH_NAME}.tfplan']
```

Comment:
this line is generating this error:

```
Finished Step #4 - "tf apply"
PUSH
Artifacts will be uploaded to gs://cft-cloudbuild-artifacts-e64e using gsutil cp
cloudbuild-tf-apply.yaml: Uploading path....
Copying file://cloudbuild-tf-apply.yaml [Content-Type=application/octet-stream]...
/ [0/1 files][    0.0 B/  1.6 KiB]   0% Done                                    
/ [1/1 files][  1.6 KiB/  1.6 KiB] 100% Done                                    
Operation completed over 1 objects/1.6 KiB.                                      
cloudbuild-tf-apply.yaml: 1 matching files uploaded
envs/prod/prod.tfplan: Uploading path....
CommandException: No URLs matched: envs/prod/prod.tfplan
CommandException: 1 file/object could not be transferred.
ERROR
ERROR: could not upload envs/prod/prod.tfplan to gs://cft-cloudbuild-artifacts-e64e/terraform/cloudbuild/apply/e24b0c50-2a0c-489a-b0b4-f26f0dc403b1/; err = gsutil exited with non-zero status: 1
```
when for example running it on step `1-org` in branch `prod` because the name actual path is `envs/shared/shared.tfplan`  not `envs/prod/prod.tfplan`

Reply:
The problem is there's no logic within the yaml for "if prod, actually do shared" - this is why I was worried it would be cleaner to keep 'shared' as a separate branch or even a separate repo with different build triggers.  

Alternatively, is there any actual benefit to uploading the plan file to the cloudbuild artifacts bucket?

Reply:
In step `3-networks ` the case is a bit more complex. 

Instead  of 
"if `prod`, actually do `shared`"
 we have 
"if `prod`, actually do `shared` **AND** `prod`".
So that we get two artifacts in one single execution.

Reply:
in step `3-networks`:
- `plan` is not able to generate plans for `dev/nonprod/prod` because they depend on resources from `shared` and it is not able to upload `shared.tfplan` 
- `apply prod` apply both `shared` and `prod` but does not upload the plan for `shared` and does not generate/upload the plan for `prod`(dependency with `shared`)
- after `apply prod`, both `apply dev` `apply nonprod` run successfully.